exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/map/UncheckedFlatMapProof.java

```diff
@@ -250,22 +256,35 @@ private CheckedMapProof checkProof() {
   }
 
   private static HashCode getEmptyProofListHash() {
-    return HashCode.fromBytes(new byte[Hashing.DEFAULT_HASH_SIZE_BYTES]);
+    return HASH_FUNCTION.newHasher()
+        .putByte(MAP_NODE_PREFIX)
+        .putBytes(new byte[Hashing.DEFAULT_HASH_SIZE_BYTES])
+        .hash();
   }
 
   private static HashCode getSingleEntryRootHash(MapProofEntry entry) {
-    return getSingleEntryRootHash(entry.getDbKey(), entry.getHash());
+    HashCode entryRootHash = getSingleEntryRootHash(entry.getDbKey(), entry.getHash());
+    return getMapProofRootHash(entryRootHash);
   }
 
   private static HashCode getSingleEntryRootHash(MapEntry<ByteString, ByteString> entry) {
     DbKey dbKey = DbKey.newLeafKey(entry.getKey());
     HashCode valueHash = HASH_FUNCTION.hashByteString(entry.getValue());
-    return getSingleEntryRootHash(dbKey, valueHash);
+    HashCode entryRootHash = getSingleEntryRootHash(dbKey, valueHash);
+    return getMapProofRootHash(entryRootHash);
+  }
+
+  private static HashCode getMapProofRootHash(HashCode merkleRoot) {
+    return HASH_FUNCTION.newHasher()
+        .putByte(MAP_NODE_PREFIX)
+        .putObject(merkleRoot, hashCodeFunnel())
+        .hash();
   }
 
   private static HashCode getSingleEntryRootHash(DbKey key, HashCode valueHash) {
     assert key.getNodeType() == Type.LEAF;
     return HASH_FUNCTION.newHasher()
+        .putByte(MAP_BRANCH_NODE_PREFIX)
```

Comment:
Is it correct (= a leaf node having a **branch** prefix)?

Reply:
No, as we discussed in wiki, for leaf we use **blob** prefix.

Reply:
@pavel-mukhanov @MakarovS Is the wiki currently wrong or this code? Shall leafNodeHash be:

(1) as this code does:
```
valueHash = sha256(BLOB || value)
leafNodeHash = sha256(BRANCH || key || valueHash)
```

or (2) as the comment and wiki suggest:
```
leafNodeHash = sha256(BLOB || value)
```

Reply:
Sorry, I'm confused you, **leaf** value should be hashed with blob prefix, **leaf node** with branch prefix. The code is right. I should change wiki once again.

Reply:
Good, thanks!

Reply:
@MakarovS , @pavel-mukhanov , 

Currently the code seems to compute the leafNodeHash as (2), is that correct? See https://github.com/exonum/exonum-java-binding/blob/f9f15066923a644ebcd00c3056abc45d0264f8df/exonum-java-binding/common/src/main/java/com/exonum/binding/common/proofs/map/UncheckedFlatMapProof.java#L231

Reply:
It seems correct, according to original code - https://github.com/exonum/exonum/blob/master/components/merkledb/src/proof_map_index/proof.rs#L561
@pavel-mukhanov please correct me if I'm wrong

Reply:
It's correct.